### PR TITLE
Add CNI version skew integration test.

### PR DESCRIPTION
**Pull request by bianpengyuan**
Fix https://github.com/istio/istio/issues/32298

Part of CNI beta graduation requirement: https://docs.google.com/document/d/1RMGiH-xFmmnct7bkBgmESQ4FONNuZ1J1V4oWQ8aB5qk/edit

This test will only run as part of postsubmit. I test ran it at presubmit by commenting out postsubmit label, and here is the result: https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34202/integ-cni-k8s-tests_istio/1418065931690577920

### CODE REVIEW

**Review by Monkeyanator — COMMENTED** [review]
looks great overall! Just a few comments

File: tests/integration/pilot/cni/cniversionskew_test.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+)
+
+// Currently only test CNI with one version behind.
+var versions = []string{NMinusOne}
+
+// TestCNIVersionSkew runs all traffic tests with older versions of CNI and lastest Istio.
+// This is to simulate the case where CNI and Istio control plane versions are out of sync during upgrade.
+func TestCNIVersionSkew(t *testing.T) {
+	framework.
+		NewTest(t).
```
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply tests/integration/pilot/cni/cniversionskew_test.go:54]: Yeah this test can run with multi clusters.

File: tests/integration/pilot/cni/cniversionskew_test.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+				}
+				common.RunAllTrafficTests(t, i, apps)
+			}
+		})
+}
+
+func TestMain(m *testing.M) {
+	framework.
+		NewSuite(m).
+		Label(label.Postsubmit).
```
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply tests/integration/pilot/cni/cniversionskew_test.go]: Done.

File: tests/integration/pilot/cni/cniversionskew_test.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+				// Make sure CNI pod is ready
+				_, err := kube.WaitUntilPodsAreReady(podFetchFn)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				// Restart all apps so that newly deployed CNI could configure IPTables for them.
+				for _, app := range apps.All {
+					app.Restart()
```
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply tests/integration/pilot/cni/cniversionskew_test.go]: Done.

File: tests/integration/pilot/cni/cniversionskew_test.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		NewSuite(m).
+		Label(label.Postsubmit).
+		Setup(istio.Setup(&i, nil)).
+		Setup(func(t resource.Context) error {
+			return common.SetupApps(t, i, apps)
+		}).
+		Run()
+}
+
+func installCNIOrFail(t framework.TestContext, ver string) {
```
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply tests/integration/pilot/cni/cniversionskew_test.go:106]: Added a comment.

File: tests/integration/pilot/testdata/upgrade/README.md [hunk 5]
```diff
@@ -32,3 +33,19 @@ spec:
 ```
 
 1. Run `tar cf 1.x.y-install.yaml.tar 1.x.y-install.yaml`
+
+1. Generate a new CNI Daemonset manifest with the following settings:
```
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply tests/integration/pilot/testdata/upgrade/README.md:37]: Yeah, sound right to be automated based on releases. I can create an issue to track this if there isn't already one.

**Comment by bianpengyuan** [comment]
Last passed run: https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34202/integ-cni-k8s-tests_istio/1418345376619958272

File: tests/integration/pilot/testdata/upgrade/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply tests/integration/pilot/testdata/upgrade/README.md:37]: Created https://github.com/istio/istio/issues/34276 to track this.

**Comment by bianpengyuan** [comment]
@Monkeyanator gentle ping, any more comments? Thanks!

**Review by Monkeyanator — APPROVED** [review]
LGTM!

**Comment by bianpengyuan** [comment]
@howardjohn @stevenctl please take a look, thanks!

**Review by howardjohn — APPROVED** [review]

**Pull request merged**